pkg/kv/kvclient/kvcoord/dist_sender.go

```diff
@@ -596,6 +639,113 @@ func makeDistSenderRangeFeedMetrics() DistSenderRangeFeedMetrics {
+			metaDistSenderKVInterceptorEstimatedReplicationBytes, labels...,
+		),
+	}
+	m.mu.pathMetrics = make(map[string]*networkPathMetrics)
+	return m
+}
+
+// MetricStruct implements the metrics.Struct interface.
+func (m *DistSenderKVInterceptorMetrics) MetricStruct() {}
+
+// updateEstimatedReplicationBytes updates the EstimatedReplicationBytes metric
+// that represents the estimated replication bytes for the given KV replication
+// network path.
+func (m *DistSenderKVInterceptorMetrics) updateEstimatedReplicationBytes(
+	fromNodeID roachpb.NodeID,
+	fromLocality roachpb.Locality,
+	toNodeID roachpb.NodeID,
+	toLocality roachpb.Locality,
+	inc int64,
+) {
+	m.getEstimatedReplicationBytes(fromNodeID, fromLocality, toNodeID, toLocality).Inc(inc)
+}
+
+// getEstimatedReplicationBytes returns the metric that represents the estimated
+// replication bytes for the given network path.
+func (m *DistSenderKVInterceptorMetrics) getEstimatedReplicationBytes(
+	fromNodeID roachpb.NodeID,
+	fromLocality roachpb.Locality,
+	toNodeID roachpb.NodeID,
+	toLocality roachpb.Locality,
+) *aggmetric.Counter {
+	// Check if we have cached values.
+	np := networkPath{fromNodeID: fromNodeID, toNodeID: toNodeID}
+	if cached, ok := m.cachedPathMetrics.Load(np); ok {
+		return (cached.(*networkPathMetrics)).EstimatedReplicationBytes
+	}
+
+	pm := m.ensureNetworkPathMetrics(fromLocality, toLocality)
+
+	// pm will never be nil, so it's fine if multiple Store calls end up racing.
+	m.cachedPathMetrics.Store(np, pm)
+
+	return pm.EstimatedReplicationBytes
+}
+
+// ensureNetworkPathMetrics ensures that the networkPathMetrics struct for the
+// given locality (from, to) pair has been initialized.
+func (m *DistSenderKVInterceptorMetrics) ensureNetworkPathMetrics(
+	fromLocality, toLocality roachpb.Locality,
+) *networkPathMetrics {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Check if we've already initialized the metric. This is the case if a
+	// different network path results in the same normalized locality values.
+	labelValues := makeLocalityLabelValues(m.baseLocality, fromLocality, toLocality)
+	storeKey := strings.Join(labelValues, ",")
+	pm, ok := m.mu.pathMetrics[storeKey]
+	if !ok {
+		pm = &networkPathMetrics{
+			EstimatedReplicationBytes: m.EstimatedReplicationBytes.AddChild(labelValues...),
+		}
+		m.mu.pathMetrics[storeKey] = pm
+	}
+	return pm
+}
+
+// makeLocalityLabelValues returns a list of label values which can be used for
+// network path metrics. This applies a mapping approach where all the values in
+// "from" and "to" localities are taken up until the first value that differs.
+// Remaining values will be padded as empty strings. Note that the keys must
+// exist in the given base locality. See test cases for more information.
+func makeLocalityLabelValues(baseLocality, fromLocality, toLocality roachpb.Locality) []string {
+	labelValues := make([]string, 2*len(baseLocality.Tiers))
+	length := int(math.Min(
```

Comment:
As of golang 1.21 there is a `min` built in that you can use here and it doesn't requiring the int/float casts.

https://pkg.go.dev/builtin#min

Reply:
Done.